File: test/language/sql_tables/cpp/BlobFieldWithChildrenInitializationTableTest.cpp
#include <cstdio>
#include <fstream>

#include "gtest/gtest.h"

#include "sql_tables/TestDb.h"

#include "zserio/RebindAlloc.h"

namespace sql_tables
{
namespace blob_field_with_children_initialization_table
{

using allocator_type = TestDb::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class BlobFieldWithChildrenInitializationTableTest : public ::testing::Test
{
public:
    BlobFieldWithChildrenInitializationTableTest()
    {
        std::remove(DB_FILE_NAME);

        m_database = new sql_tables::TestDb(DB_FILE_NAME);
        m_database->createSchema();
    }

    ~BlobFieldWithChildrenInitializationTableTest() override
    {
        delete m_database;
    }

protected:
    static void fillRow(BlobFieldWithChildrenInitializationTable::Row& row, size_t index)
    {
        row.setId(static_cast<uint32_t>(index));
        const uint32_t arrayLength = static_cast<uint32_t>(index);
        vector_type<uint32_t> array;
        for (uint32_t i = 0; i < arrayLength; ++i)
            array.push_back(i);
        row.setBlob(BlobWithChildrenInitialization{arrayLength, ParameterizedArray{std::move(array)}});
    }

    static void fillRows(vector_type<BlobFieldWithChildrenInitializationTable::Row>& rows)
    {
        rows.clear();
        for (size_t i = 0; i < NUM_ROWS; ++i)
        {
            BlobFieldWithChildrenInitializationTable::Row row;
            fillRow(row, i);
            rows.push_back(row);
        }
    }

    static void checkRow(const BlobFieldWithChildrenInitializationTable::Row& row1,
            const BlobFieldWithChildrenInitializationTable::Row& row2)
    {
        ASSERT_EQ(row1.getId(), row2.getId());
        ASSERT_EQ(row1.getBlob(), row2.getBlob());
    }

    static void checkRows(const vector_type<BlobFieldWithChildrenInitializationTable::Row>& rows1,
            const vector_type<BlobFieldWithChildrenInitializationTable::Row>& rows2)
    {
        ASSERT_EQ(rows1.size(), rows2.size());
        for (size_t i = 0; i < rows1.size(); ++i)
            checkRow(rows1[i], rows2[i]);
    }

    static const char DB_FILE_NAME[];
    static const size_t NUM_ROWS;

    sql_tables::TestDb* m_database;
};

const char BlobFieldWithChildrenInitializationTableTest::DB_FILE_NAME[] =
        "language/sql_tables/blob_field_with_children_initialization_table_test.sqlite";
const size_t BlobFieldWithChildrenInitializationTableTest::NUM_ROWS = 5;

TEST_F(BlobFieldWithChildrenInitializationTableTest, readWithoutCondition)
{
    BlobFieldWithChildrenInitializationTable& table = m_database->getBlobFieldWithChildrenInitializationTable();

    vector_type<BlobFieldWithChildrenInitializationTable::Row> rows;
    fillRows(rows);
    table.write(rows);

    vector_type<BlobFieldWithChildrenInitializationTable::Row> readRows;
    auto reader = table.createReader();
    while (reader.hasNext())
        readRows.push_back(reader.next());

    checkRows(rows, readRows);
}

} // namespace blob_field_with_children_initialization_table
} // namespace sql_tables

File: test/extensions/language/sql_tables/java/sql_tables/blob_field_with_children_initialization_table/BlobFieldWithChildrenInitializationTableTest.java
package sql_tables.blob_field_with_children_initialization_table;

import static org.junit.jupiter.api.Assertions.*;

import java.io.File;
import java.io.IOException;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import zserio.runtime.ZserioError;

import sql_tables.TestDb;
import test_utils.FileUtil;
import test_utils.JdbcUtil;

public class BlobFieldWithChildrenInitializationTableTest
{
    @BeforeAll
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @BeforeEach
    public void setUp() throws IOException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new TestDb(file.toString());
        database.createSchema();
    }

    @AfterEach
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void readWithoutCondition() throws SQLException, IOException, ZserioError
    {
        final BlobFieldWithChildrenInitializationTable table =
                database.getBlobFieldWithChildrenInitializationTable();

        final List<BlobFieldWithChildrenInitializationTableRow> writtenRows =
                new ArrayList<BlobFieldWithChildrenInitializationTableRow>();
        fillRows(writtenRows);
        table.write(writtenRows);

        final List<BlobFieldWithChildrenInitializationTableRow> readRows = table.read();
        checkRows(writtenRows, readRows);
    }

    private static void fillRows(List<BlobFieldWithChildrenInitializationTableRow> rows)
    {
        for (int i = 0; i < NUM_ROWS; ++i)
        {
            rows.add(createRow(i));
        }
    }

    private static BlobFieldWithChildrenInitializationTableRow createRow(int index)
    {
        final BlobFieldWithChildrenInitializationTableRow row =
                new BlobFieldWithChildrenInitializationTableRow();
        row.setId(index);
        final int arrayLength = index;

        final long[] array = new long[arrayLength];
        for (int i = 0; i < arrayLength; ++i)
            array[i] = i;
        ParameterizedArray parameterizedArray = new ParameterizedArray(arrayLength, array);
        BlobWithChildrenInitialization blob =
                new BlobWithChildrenInitialization(arrayLength, parameterizedArray);
        row.setBlob(blob);

        return row;
    }

    private static void checkRows(List<BlobFieldWithChildrenInitializationTableRow> rows1,
            List<BlobFieldWithChildrenInitializationTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
            checkRow(rows1.get(i), rows2.get(i));
    }

    private static void checkRow(
            BlobFieldWithChildrenInitializationTableRow row1, BlobFieldWithChildrenInitializationTableRow row2)
    {
        assertEquals(row1.getId(), row2.getId());
        assertEquals(row1.getBlob(), row2.getBlob());
    }

    private static final int NUM_ROWS = 5;
    private static final String FILE_NAME = "blob_field_with_children_initialization_table_test.sqlite";

    private final File file = new File(FILE_NAME);
    private TestDb database = null;
}
